### PR TITLE
RETINA ViT-B/16 SNGP - ML Collections Config Compatibility with W&B

### PR DESCRIPTION
**Pull request by nband**
Note: for Weights and Biases compatibility, we create a joint argument for total and warmup steps. After W&B tuning for the upcoming benchmark, we will separate these arguments to be consistent with other parts of uncertainty-baselines.

### CODE REVIEW

File: baselines/diabetic_retinopathy_detection/jax_finetune_deterministic.py [hunk 1]
```diff
@@ -14,147 +14,89 @@
...
+tf.config.experimental.set_visible_devices([], 'TPU')
+
 import functools
 import itertools
 import multiprocessing
 import numbers
 import os
 import time
 
+import flax
```
Comment (dustinvtran) [inline baselines/diabetic_retinopathy_detection/jax_finetune_deterministic.py:30]: Can we revert the import reordering? When merging #735, the import order change failed the linter.

https://github.com/google/uncertainty-baselines/commit/62843fb31276d8020adda6d295edeab71ffa0b4b

File: baselines/diabetic_retinopathy_detection/experiments/config/imagenet21k_vit_base16_finetune.py [hunk 2]
```diff
@@ -87,20 +90,54 @@ def get_config():
...
+  # Input resolution of each retina image. (Default: 512)
+  config.pp_input_res = 512  # pylint: disable=invalid-name
+  pp_common = f'|onehot({config.num_classes})'
+  config.pp_train = (
+      f'diabetic_retinopathy_preprocess({config.pp_input_res})' + pp_common)
+  config.pp_eval = (
+      f'diabetic_retinopathy_preprocess({config.pp_input_res})' + pp_common)
+
+  # Training Misc
+  config.batch_size = 512  # using TPUv3-64
```
Comment (dustinvtran) [inline baselines/diabetic_retinopathy_detection/experiments/config/imagenet21k_vit_base16_finetune.py:104]: Following discussion, let's revert this to `128  # using TPUv3-8`

File: baselines/diabetic_retinopathy_detection/experiments/config/imagenet21k_vit_base16_finetune.py [hunk 3]
```diff
@@ -87,20 +90,54 @@ def get_config():
...
+
+  # Evaluation Misc
+  config.only_eval = False  # Disables training, only evaluates the model
+  config.use_validation = True  # Whether to use a validation split
+  config.use_test = True  # Whether to use a test split
+
+  # Step Counts
+
+  # Varied together for wandb sweep compatibility.
+  config.total_and_warmup_steps = (10_000, 500)
```
Comment (dustinvtran) [inline baselines/diabetic_retinopathy_detection/experiments/config/imagenet21k_vit_base16_finetune.py:134]: Does wandb support a zip-type iterator instead of just the product of all possible sweeps? It's preferable to keep these two hparams as two config settings instead of a tuple.

File: baselines/diabetic_retinopathy_detection/jax_finetune_deterministic.py [hunk 4]
```diff
@@ -14,147 +14,89 @@
...
-
-# Evaluation flags.
-flags.DEFINE_bool('only_eval', False,
-                  'Disables training, only evaluates the model.')
-flags.DEFINE_bool('use_validation', True,
-                  'Whether to use a validation split.')
-flags.DEFINE_bool('use_test', True, 'Whether to use a test split.')
+logging.info(tf.config.experimental.get_visible_devices())
 
+ml_collections.config_flags.DEFINE_config_file(
```
Comment (dustinvtran) [inline baselines/diabetic_retinopathy_detection/jax_finetune_deterministic.py:58]: Following above comment about `total_steps` and `warmup_steps`, let's keep the config locked.

File: baselines/diabetic_retinopathy_detection/utils/vit_utils.py [hunk 5]
```diff
@@ -40,13 +40,6 @@
 from . import results_storage_utils  # local file import
 import wandb
```
Comment (dustinvtran) [inline baselines/diabetic_retinopathy_detection/utils/vit_utils.py:42]: These changes should be reflected now in the main branch. Can you rebase?

**Review by dustinvtran — CHANGES_REQUESTED** [review]
This looks great! Thanks Neil. Added minor comments.

File: baselines/diabetic_retinopathy_detection/jax_finetune_deterministic.py [hunk 6]
```diff
@@ -128,11 +129,7 @@ def main(argv):
   # The pool is used to perform misc operations such as logging in async way.
   pool = multiprocessing.pool.ThreadPool()
 
-  def write_note(note):
-    if jax.process_index() == 0:
-      logging.info('NOTE: %s', note)
-
-  write_note('Initializing...')
+  vit_utils.write_note('Initializing...')
```
Comment (dustinvtran) [inline baselines/diabetic_retinopathy_detection/jax_finetune_deterministic.py:132]: Can we keep the write_note defined in the script? Internally we actually modify `write_note` to write additional logs, and the function's scoped to use an outside variable.  If we move the function to a separate module, we'd have to augment `write_note` to pass a second argument.

File: baselines/diabetic_retinopathy_detection/experiments/config/imagenet21k_vit_base16_finetune.py (hunk 3 above)
Comment (dustinvtran) [reply baselines/diabetic_retinopathy_detection/experiments/config/imagenet21k_vit_base16_finetune.py:134]: Following discussion, let's add a TODO here.

**Review by dustinvtran — CHANGES_REQUESTED** [review]

File: baselines/diabetic_retinopathy_detection/jax_finetune_deterministic.py (hunk 6 above)
Comment (nband) [reply baselines/diabetic_retinopathy_detection/jax_finetune_deterministic.py:134]: Sure thing -- wasn't aware of the internal changes!

File: baselines/diabetic_retinopathy_detection/jax_finetune_deterministic.py [hunk 7]
```diff
@@ -397,7 +400,8 @@ def decay_fn(v, wd):
   # Advance the iterators if we are restarting from an earlier checkpoint.
   # TODO(dusenberrymw): Look into checkpointing dataset state instead.
   if first_step > 0:
-    write_note('Advancing iterators after resuming from a checkpoint...')
+    write_note(
```
Comment (dustinvtran) [inline baselines/diabetic_retinopathy_detection/jax_finetune_deterministic.py:403]: nit: revert newline here

File: baselines/diabetic_retinopathy_detection/jax_finetune_deterministic.py [hunk 8]
```diff
@@ -515,7 +519,7 @@ def decay_fn(v, wd):
         results_arrs['y_true'] = np.concatenate(results_arrs['y_true'], axis=0)
         results_arrs['y_pred'] = np.concatenate(
             results_arrs['y_pred'], axis=0).astype('float64')
-        results_arrs['y_pred_entropy'] = vit_utils.entropy(
+        results_arrs['y_pred_entropy'] = entropy(
```
Comment (dustinvtran) [inline baselines/diabetic_retinopathy_detection/jax_finetune_deterministic.py:522]: Let's keep with the call to vit_utils. Internally, I had to do this because SciPy uses an older version internally that doesn't support the `axis` argument, so I copied the function.

File: baselines/diabetic_retinopathy_detection/jax_finetune_deterministic.py [hunk 9]
```diff
@@ -382,7 +385,7 @@ def decay_fn(v, wd):
       map(lr_fn, range(total_steps)), config.get('prefetch_to_device', 1))
 
   write_note(f'Replicating...\n{chrono.note}')
-  opt_repl = flax.jax_utils.replicate(opt_cpu)
+  opt_repl = flax_utils.replicate(opt_cpu)
```
Comment (dustinvtran) [inline baselines/diabetic_retinopathy_detection/jax_finetune_deterministic.py:388]: nit: revert to flax.jax_utils to avoid polluting the namespace

**Review by dustinvtran — CHANGES_REQUESTED** [review]

**Review by dustinvtran — APPROVED** [review]
